src/mlpack/bindings/R/default_param_impl.hpp

```diff
@@ -0,0 +1,145 @@
+
+#include "default_param.hpp"
+
+namespace mlpack {
+namespace bindings {
+namespace r {
+
+/**
+ * Return the default value of an option.
+ */
+template<typename T>
+std::string DefaultParamImpl(
+    util::ParamData& data,
+    const typename boost::disable_if<arma::is_arma_type<T>>::type* /* junk */,
+    const typename boost::disable_if<util::IsStdVector<T>>::type* /* junk */,
+    const typename boost::disable_if<data::HasSerialize<T>>::type* /* junk */,
+    const typename boost::disable_if<std::is_same<T, std::string>>::type*,
+    const typename boost::disable_if<std::is_same<T,
+        std::tuple<mlpack::data::DatasetInfo, arma::mat>>>::type* /* junk */)
+{
+  std::ostringstream oss;
+  if (std::is_same<T, bool>::value)
+    oss << "FALSE";
+  else
+    oss << boost::any_cast<T>(data.value);
+
+  return oss.str();
+}
+
+/**
+ * Return the default value of a vector option.
+ */
+template<typename T>
+std::string DefaultParamImpl(
+    util::ParamData& data,
+    const typename boost::enable_if<util::IsStdVector<T>>::type* /* junk */)
+{
+  // Print each element in an array delimited by square brackets.
+  std::ostringstream oss;
+  const T& vector = boost::any_cast<T>(data.value);
+  oss << "c(";
+  if (std::is_same<T, std::vector<std::string>>::value)
+  {
+    if (vector.size() > 0)
+    {
+      for (size_t i = 0; i < vector.size() - 1; ++i)
+      {
+        oss << "'" << vector[i] << "', ";
+      }
+
+      oss << "'" << vector[vector.size() - 1] << "'";
+    }
+
+    oss << ")";
+  }
+  else
+  {
+    if (vector.size() > 0)
+    {
+      for (size_t i = 0; i < vector.size() - 1; ++i)
+      {
+        oss << vector[i] << ", ";
+      }
+
+      oss << vector[vector.size() - 1];
+    }
+
+    oss << ")";
+  }
+  return oss.str();
+}
+
+/**
+ * Return the default value of a string option.
+ */
+template<typename T>
+std::string DefaultParamImpl(
+    util::ParamData& data,
+    const typename boost::enable_if<std::is_same<T, std::string>>::type*)
+{
+  const std::string& s = *boost::any_cast<std::string>(&data.value);
+  return "\"" + s + "\"";
+}
+
+/**
+ * Return the default value of a matrix option (this returns the default
+ * filename, or '' if the default is no file).
+ */
+template<typename T>
+std::string DefaultParamImpl(
+    util::ParamData& /* data */,
+    const typename boost::enable_if_c<
+        arma::is_arma_type<T>::value ||
+        std::is_same<T, std::tuple<mlpack::data::DatasetInfo,
+                                   arma::mat>>::value>::type* /* junk */)
+{
+  // Get the filename and return it, or return an empty string.
+  if (std::is_same<T, arma::rowvec>::value ||
+      std::is_same<T, arma::vec>::value ||
+      std::is_same<T, arma::mat>::value)
+  {
+    return "matrix()";
+  }
+  else if (std::is_same<T, arma::Row<size_t>>::value ||
+      std::is_same<T, arma::Col<size_t>>::value ||
+      std::is_same<T, arma::Mat<size_t>>::value)
+  {
+    return "matrix(as.integer())";
```

Comment:
```suggestion
    return "matrix(as.integer(c(...)))";
```

Reply:
The idea here is to provide the R code for the default for an empty matrix of integer type... I think that `c(...)` won't work since the user would have to fill out what `...` is.  Let me know if I misunderstood---I'm certainly not an expert.  But I did find that `matrix(as.integer())` worked and gave an empty matrix (I think) but `matrix(as.integer(c(...)))` failed:

```
> matrix(as.integer(c(...)))
Error in matrix(as.integer(c(...))) : '...' used in an incorrect context
```

Reply:
@rcurtin so, the reason for `c()` being present is to hint that it should be formatted as a vector first. We can drop the `...` part. 

```r
matrix(as.integer(2))
# [1] 2

matrix(as.integer(c(2, 3)))
#      [,1]
# [1,]    2
# [2,]    3
```

Reply:
I have updated this as suggested.

Reply:
Thanks @Yashwants19.  I am sure there is something I am misunderstanding here, but if I run the two "default values" for integer and numeric matrices, I get this output:

```
> matrix(as.integer(c()))
     [,1]
> matrix(c())
Error in matrix(c()) : 'data' must be of a vector type, was 'NULL'
```

It seems also like `matrix(as.integer())` gives identical output to `matrix(as.integer(c()))`:

```
> matrix(as.integer())
     [,1]
```

The default values we are printing there should be

 * an empty matrix
 * an empty matrix that holds `size_t`s or equivalent
 * an empty vector
 * an empty vector that holds `size_t`s or equivalent

But it seems like these outputs aren't empty.  Maybe there is a better way that we provide the output here?

Here's an example from Julia:
https://www.mlpack.org/doc/mlpack-git/julia_documentation.html#adaboost

Specifically the `default` column is what's being generated with this code.  The idea here is that when a matrix type *isn't* being passed as an input parameter to the binding (like, with the AdaBoost binding, `test` or `training` or `labels`), the `default` column shows what its value actually is.  If there is a concept of "nothing" in R, that could work too.

(For what it's worth, what we print for the default value of an unpassed matrix-type parameter is totally up for discussion.  Originally, I thought it best to print actual working code in the language of choice that represents an empty matrix.)

Reply:
Hi @rcurtin, I have updated the default value to `matrix(as.double(c()), 0, 0)` and `matrix(as.integer(c()), 0, 0)` which gives output:
```R
> matrix(as.double(c()), 0, 0)
<0 x 0 matrix>
> class(matrix(as.double(c()), 0, 0))
[1] "matrix" "array" 
> typeof(matrix(as.double(c()), 0, 0))
[1] "double"

> matrix(as.integer(c()), 0, 0)
<0 x 0 matrix>
> class(matrix(as.integer(c()), 0, 0))
[1] "matrix" "array" 
> typeof(matrix(as.integer(c()), 0, 0))
[1] "integer"
```
Let me know, What do you think about this.

Reply:
Awesome---to me that seems fine, if @coatless and @eddelbuettel think it's reasonable. :+1: